### ISSUE TITLE
FLAG_NOPAGING Question

### ISSUE THREAD

**Issue by LeetPro360** [ufrisk/MemProcFS#90]
Hi Ufrisk. First off, great work and thank you.

Secondly, I am using a DMA card to read memory on my target machine. I have a question, am I limited to reading only memory that's on my RAM or am I also able to read memory that has been paged out into a page file on my SSD?

I thought a DMA card only had access to the RAM so I thought it was impossible to read a paged file but I do see there is an option in vmm.memRead that has a flag FLAG_NOPAGING?

Thank you in advance

**Comment by ufrisk** [comment]
Paging on Windows is more complex than just being in RAM (marked as active in page tables). If a page entry is non-existant or marked as paged-out this may mean (amongst others):
- backed by transient page still in RAM
- backed by prototype pte page still in RAM
- backed by zero memory
- backed by compressed memory (still in RAM)
- backed by page file (paged out to disk). MemProcFS also support this if you specify it in the -pagefile option

basically "paged out" means way more than "page file".

**Comment by imerzan** [comment]
Also interested in this. So reading memory via win32 ReadProcessMemory, the OS is able to abstract this request and return results whether it's in physical memory, page file , etc. 

So if I understand correctly, MemProcFs cannot read virtual memory addresses when the contents are not in physical memory --- is that correct?

**Comment by imerzan** [comment]
I am interested in this as well. I was able to use the argument
`-pagefile pagefile.sys`

And it seems to init OK.

Does this mean if MemProcFs tries to resolve a virtual address, and it is not currently in physical memory, that it will check the pagefile instead?

**Comment by ufrisk** [comment]
If MemProcFS tries to resolve a physical address and it's "paged out" it will will try resolve this, it may end up resolving that it's compressed in RAM at another virtual address. That address in turn may resolve to a page file. These resolves may happen in multiple layers and may be slow, hence the flag. That it's "paged out" does not necessarily mean it's in the page file. It may just be that it resides still in-memory but on other places...

https://github.com/ufrisk/MemProcFS/wiki/_CommandLine for documentation

in most cases it's `-pagefile0 pagefile.sys` and ` -pagefile1 swapfile.sys`  but Windows may have up to 16 "page files".

the page file must be taken at the same time as the memory dump for it to be valid. Just a minute apart and there will be considerable drift/corruptions. But in some cases it's useful.

**Comment by ufrisk** [comment]
If you're dumping over a period of 30 minutes the page file will be a 95% unusable mess of garbage data.

**Comment by imerzan** [comment]
OK, that's what I thought. So in other words, if I need to analyze system memory (that may be "paged out to disk") over this long timespan, it would be in my best interests to disable the system page file on the target system.

**Comment by ufrisk** [comment]
yes, but only if there is a high memory pressure on the target, if there are lots of free memory it won't really page anything to file, or that much at least.

**Comment by ufrisk** [comment]
I hope I have answered these questions. If not please let me know. Closing issue.
